rust/kernel/net/device.rs

```diff
@@ -0,0 +1,1076 @@
+    /// Only invoke the rx handler of L3 master device.
+    L3MDEV_RX_HANDLER = bindings::netdev_priv_flags_IFF_L3MDEV_RX_HANDLER,
+    /// Rename is allowed while device is up and running.
+    LIVE_RENAME_OK = bindings::netdev_priv_flags_IFF_LIVE_RENAME_OK,
+}
+
+/// Feature flags.
+pub mod feature {
+    use core::convert::From;
+    use core::ops::{Add, AddAssign, Deref, DerefMut, Sub, SubAssign};
+
+    /// Holds multiple flags to give to an interface via [`super::NetDevice::add_flag`].
+    #[repr(transparent)]
+    #[derive(Debug, Clone, Copy)]
+    pub struct NetIF(u64);
+
+    impl NetIF {
+        /// Create new Flag with value `0`.
+        pub const fn new() -> Self {
+            Self(0)
+        }
+
+        /// Add flag to Self.
+        pub fn add_flag(&mut self, flag: u64) {
+            self.0 |= flag;
+        }
+
+        /// Remove the given flag from Self.
+        pub fn remove_flag(&mut self, flag: u64) {
+            self.0 &= !(flag);
+        }
+    }
+
+    impl Deref for NetIF {
+        type Target = u64;
+
+        fn deref(&self) -> &Self::Target {
+            &self.0
+        }
+    }
+
+    impl DerefMut for NetIF {
+        fn deref_mut(&mut self) -> &mut Self::Target {
+            &mut self.0
+        }
+    }
+
+    impl Add for NetIF {
+        type Output = Self;
+
+        #[allow(clippy::suspicious_arithmetic_impl)]
+        fn add(self, rhs: Self) -> Self::Output {
+            Self::from(self.0 | rhs.0)
+        }
+    }
+
+    impl Add<u64> for NetIF {
+        type Output = Self;
+
+        #[allow(clippy::suspicious_arithmetic_impl)]
+        fn add(self, rhs: u64) -> Self::Output {
+            Self::from(self.0 | rhs)
+        }
+    }
+
+    impl Sub for NetIF {
+        type Output = Self;
+
+        #[allow(clippy::suspicious_arithmetic_impl)]
+        fn sub(self, rhs: Self) -> Self::Output {
+            Self::from(self.0 & !rhs.0)
+        }
+    }
+
+    impl Sub<u64> for NetIF {
+        type Output = Self;
+
+        #[allow(clippy::suspicious_arithmetic_impl)]
+        fn sub(self, rhs: u64) -> Self::Output {
+            Self::from(self.0 & !rhs)
+        }
+    }
+
+    impl AddAssign for NetIF {
+        #[allow(clippy::suspicious_op_assign_impl)]
+        fn add_assign(&mut self, rhs: Self) {
+            self.0 |= rhs.0
+        }
+    }
+
+    impl AddAssign<u64> for NetIF {
+        #[allow(clippy::suspicious_op_assign_impl)]
+        fn add_assign(&mut self, rhs: u64) {
+            self.0 |= rhs
+        }
+    }
+
+    impl SubAssign for NetIF {
+        #[allow(clippy::suspicious_op_assign_impl)]
+        fn sub_assign(&mut self, rhs: Self) {
+            self.0 &= !rhs.0
+        }
+    }
+
+    impl SubAssign<u64> for NetIF {
+        #[allow(clippy::suspicious_op_assign_impl)]
+        fn sub_assign(&mut self, rhs: u64) {
+            self.0 &= !rhs
+        }
+    }
+
+    impl From<u64> for NetIF {
+        fn from(flags: u64) -> Self {
```

Comment:
Better to use a method rather than `From`, so it'll be easier to implement validity checks in the future. Again, refer to `bitflags` as the best practice.

Reply:
what about TryFrom?

Reply:
I think we can keep `From<NetIF> for u64, as NetIF should alway contain a valid bitmask?

Reply:
I think that conversion should better be more explicit, e.g. via a method. Driver shouldn't handle the raw values anyway.

Reply:
It is a method, it is just a method from a trait. You have to call from or into.

Reply:
I had a look at bitflags and it seems to be a pending issue: https://github.com/bitflags/bitflags/issues/224.

Reply:
Yes, what does it mean for me? Leave it with From<NetIF> and TryFrom<u64>? or something else?

Reply:
I guess leave it like that for now.

BTW, please annotate these simple method and trait impl methods with `#[inline]`, otherwise they won't be inlined in the driver.

Reply:
`#[inline]` for empty as well?

Reply:
Sprayed some inline there, @nbdd0121  could you look at it?

Reply:
Yes